Use typed-rest-client 1.0.7 to allow toolLib.downloadTool to work with redirects.

sanderaernouts opened the pull request:
# Background
While trying to download a tools exe from Github releases using toolLib.downloadTool I got the following error: UnhandledPromiseRejectionWarning: Error: Unexpected HTTP response: 302. Github was redirecting met to an AWS blob which seems perfectly fine to me but toolLib.downloadTool only accepts HTTP 200.

# Change
I updated typed-rest-client to 1.0.7 to allow toolLib.downloadTool to work with HTTP redirects.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vsts-task-tool-lib?pullRequest=28) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vsts-task-tool-lib?pullRequest=28) before we can accept your contribution.<br/><br/>:x: sanderaernouts [sign now](https://cla.opensource.microsoft.com/Microsoft/vsts-task-tool-lib?pullRequest=28)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vsts-task-tool-lib?pullRequest=28) it.</sub>

stephenmichaelf commented:
@sanderaernouts Thanks for taking the time to do a PR, we appreciate it! This module relies on the typed-rest-client which we also own. We recently made changes [here](https://github.com/Microsoft/typed-rest-client/blob/dffabe6a69a456c1672dae7e1b86d7a37f9017f6/lib/HttpClient.ts#L212) to have that client handle redirects. It's a lower level client that should have that responsibility. I think if we simply update this module to point to the latest typed-rest-client it should resolve your issue.

Since you already started working on this and investigating do you want to give that a try and see if it works?

sanderaernouts commented:
@stephenmichaelf  sure thing. I updated typed-rest-client to 1.0.7 and added some test using https://httpbin.org to test the redirect status codes work with downloadTool.

sanderaernouts commented:
@stephenmichaelf sure thing. I updated typed-rest-client to 1.0.7 and added some tests to verify the redirect status codes work with downloadTool. I did not change the statuscode back to statuscode == 200 because there are other 2xx status codes that indicate success.

bryanmacfarlane commented on tool.ts:
```diff
@@ -222,7 +222,7 @@ export async function downloadTool(url: string, fileName?: string): Promise<stri
             tl.debug('downloading');
             let response: httpm.HttpClientResponse = await http.get(url);
-            if (response.message.statusCode != 200) {
+            if (response.message.statusCode >= 400) {
```
We don't need this if typed rest client is handling 3xx.  if there's an large / infinite redirect issue, typed rest client will appropriately error and this shouldn't mask it.  Same for 500s.

bryanmacfarlane commented on package.json:
```diff
@@ -25,7 +25,7 @@
     "@types/semver": "^5.3.0",
     "semver-compare": "^1.0.0",
-    "typed-rest-client": "0.14.1",
+    "typed-rest-client": "1.0.7",
```
This is good.

bryanmacfarlane commented on .gitignore:
```diff
@@ -34,8 +34,17 @@ build/Release
 jspm_packages
 
+# Npm lock files
+package-lock.json
```
this gets checked in and it shouldn't be ignored.  If this changes you should see the diff

bryanmacfarlane commented on test/toolTests.ts:
```diff
@@ -81,24 +81,83 @@ describe('Tool Tests', function () {
-                let nonExistentUrl: string = "http://httpbin.org/nonexistent";
-                let downPath: string = await toolLib.downloadTool(nonExistentUrl);
+    
+    describe('has status code in exception dictionary for HTTP error code responses', function () {
```
I think the only test we need is a download that redirects (once) and we make sure it's 200.  The tests below are the responsibility of the typed rest client

stephenmichaelf commented on .gitignore:
```diff
@@ -34,8 +34,17 @@ build/Release
+TEMP
+
+# Resources
+Strings/*
```
We need these files.

sanderaernouts replied:
Thought I removed that one >..<

sanderaernouts replied:
Does TEMP need to be removed from the .gitignore as well?

stephenmichaelf commented on test/toolTests.ts:
```diff
@@ -81,24 +81,83 @@ describe('Tool Tests', function () {
+        
     });
 
+    describe('accepts any status code < 400', function() {
```
Can remove this test.

stephenmichaelf replied:
Or modify to what Bryan said, check that for a redirect URL we get the final URL and a 200.

stephenmichaelf commented:
Thanks for taking the time to make these adjustments, it is much appreciated.

LGTM. 

@bryanmacfarlane if you agree I am fine merging this.

stephenmichaelf commented:
Bump @bryanmacfarlane.

bryanmacfarlane commented:
Thanks!

bryanmacfarlane commented:
I will release after lunch :)

bryanmacfarlane commented:
I published 0.5.1

stephenmichaelf commented:
Thanks!
